sequence_processing_pipeline/Pipeline.py

```diff
@@ -276,6 +297,55 @@ def _validate_sample_sheet(self, sample_sheet_path):
                               isinstance(x, WarningMessage)]
             return val_sheet
 
+    def _validate_mapping_file(self, mapping_file_path):
+        """
+        Performs validation for mapping-files.
+        :return: If successful, an empty list of strings and a valid
+                 mapping-file. If unsuccessful, a list of warning and error
+                 messages and None.
+        """
+        try:
+            df = pd.read_csv(mapping_file_path, delimiter='\t', dtype=str)
+        except pd.errors.ParserError:
+            # ignore parser errors as they obviously prove this is not a
+            # valid mapping file. Return None w/error message
+            raise PipelineError('Cannot parse mapping-file.')
+
+        dupes = Pipeline._identify_duplicate_columns(df.columns)
```

Comment:
what's the difference between this method and the obs/exp code?

Reply:
The logic in identify_duplicate_columns() code was added in response to having a number of jobs fail late into processing when seqpro identified duplicate columns in the mapping file. It's job here is to identify duplicate columns to the user and abort early.

The obs/exp code is completely different in that once we've confirmed that all of the columns are unique, we need to confirm that all of the required columns exist in the mapping file. We also want to give the user a warning when it's found columns it wasn't expecting to see, like a newly created column or a misspelled column name.

I've gone ahead and folded the code in _identify_duplicate_columns() into this function, since it's only used in one place, and added some additional comments. This should clear up any future confusion.